src/CmsChangeLogBundle/src/esono/pkgcmschangelog/TCMSListManager/TCMSListManagerFieldHistory.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+use ChameleonSystem\CoreBundle\ServiceLocator;
+use ChameleonSystem\CoreBundle\Translation\ChameleonTranslator;
+use Doctrine\DBAL\Connection;
+
+class TCMSListManagerFieldHistory extends TCMSListManagerFullGroupTable
+{
+
+    /**
+     * {@inheritDoc}
+     */
+    protected function DefineInterface(): void
+    {
+        $this->methodCallAllowed = ['restoreFieldValueVersion'];
+    }
+
+    public function restoreFieldValueVersion(): void
+    {
+        // TODO: Implement service or inline functionality to restore selected revision.
+        // Selected id is supplied at "recordId" parameter in request body.
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function AddFields(): void
+    {
+        // Only fields necessary for a comprehensive listing of field value changes are defined, default fields from parent are explicitly excluded.
+
+        $translator = $this->getTranslator();
+        $linkField = ['id'];
+
+        ++$this->columnCount;
+        $this->tableObj->AddHeaderField($translator->trans('chameleon_system_cms_change_log.column.change_type'), 'left', null, 1, false);
+        $this->tableObj->AddColumn('pkg_cms_changelog_set_change_type', 'left', null, $linkField);
+
+        ++$this->columnCount;
+        $this->tableObj->AddHeaderField($translator->trans('chameleon_system_cms_change_log.column.changed_on'), 'left', null, 1, false);
+        $this->tableObj->AddColumn('pkg_cms_changelog_set_modify_date', 'left', null, $linkField);
+
+        ++$this->columnCount;
+        $this->tableObj->AddHeaderField($translator->trans('chameleon_system_cms_change_log.column.old_value'), 'left', null, 1, false);
+        $this->tableObj->AddColumn('value_old', 'left', [$this, 'getFieldTextWithAttributes'], $linkField);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function GetCustomRestriction(): string
+    {
+        $changeLogFieldConfigurationId = $this->sRestriction;
+
+        return sprintf(
+            '`pkg_cms_changelog_item`.`cms_field_conf` = %s',
+            $this->getDatabaseConnection()->quote($changeLogFieldConfigurationId)
+        );
+    }
+
+    /**
+     * @return string
```

Comment:
What about exception then? A keyword for exceptions in the method signature like...

```swift
func performOperation() throws -> Result { ... }
```

...probably isn't on the horizon, so it should be clear what *should* be part of the docblock with parameter types and return types. What about:

1. Parameters with the non-existent `any` type
2. Return types neither matching scalar/type nor interface (`: any`)
3. Exceptions thrown from the method

My frontmost question is: If a doc block still needs to be used for _some_ parts of functions but not for others, does an incomplete docblock not lead to problems with comprehensiveness? Should it only be removed for absolutely obvious cases like the one above? What about `{@inheritDoc}`?

Reply:
If there is any information that can't be expressed in the signature itself (mostly because of exception declaration), a complete docblock should be written, even if that means duplicate declaration.

I think you mean `mixed` instead of `any` - this type should be avoided. In almost all cases we're better off with declaring concrete types. If we ever use `mixed`, a docblock will also not add any further information and can be omitted.

In case I didn't miss exceptions thrown by this method, we should remove the docblock.

Reply:
Alright, that's easy to remember then, I skimmed through.